Separate device attestation and operational cert provisioning into separate steps

cecille opened the pull request:
#### Problem
Device attestation and operational certificate provisioning are currently implemented as a long chain of cluster commands in the device commissioner. While this works well currently, it means that all the information required for attestation and op cert provisioning needs to be available when the command chain is triggered. it is somewhat limiting for commissioners that have more complex certificate provisioning requirements where they may need to use the information provided by earlier steps, then perform external steps to retrieve or generate certs before continuing. This PR breaks these long chains into individual commissioning steps with callbacks and drives the chain from the auto commissioner class. This means that a commissioner could replace the auto commissioner delegate with a delegate that performs additional work between the phases of the op cert provisioning. This will also allow us to do a bit of code cleanup to handle the error cases in a more centralized way (later PR).

#### Change overview
Separates the device attestation and certification provisioning command chains into individual commissioning steps and drives these from the external autocommissioner class. 
NOTE: There was a fair amount of code moved in this change. The changes are broken up into the individual steps as separate commits to make the review easier to parse.

#### Testing
BLE testing: chip-tool with M5 all clusters app.
IP testing: linux lighting app with chip-tool.
also covered by cirque tests.

cecille committed:
DAC and PAI storage/control to AutoCommissioner.

cecille committed:
Move attestation step to commissioning delegate

cecille committed:
Move signing request and NOC chain gen to AutoCommissioner.

cecille committed:
Remaining steps pulled out into external commissioner.

cecille committed:
Fix up error handling

cecille commented:
Putting this back to draft for a minute while I work out the conflicts.

cecille commented:
Oop, not as bad as I thought it would be.

cecille committed:
Add some additional logging

I can't repro the cirque failure locally for some reason. Adding
logging to figure out what's up.

msandstedt commented on src/controller/CommissioningDelegate.h:
```diff
@@ -105,17 +149,33 @@ class CommissioningDelegate
+        CommissioningReport(CommissioningStage stage) : stageCompleted(stage) {}
         CommissioningStage stageCompleted;
         // TODO: Add other things the delegate needs to know.
         union
         {
```
We have a c++14-compatible variant that can do this in a type-safe manner.  See src/lib/support/Variant.h.

cecille replied:
Didn't know about that - thanks. Would you mind if I made an issue and addressed this in a follow up PR? It would keep the change for that more contained.

msandstedt replied:
Yes, that's fine.

cecille replied:
Actually, never mind - had to wait for CI so I just changed it here. Wasn't as much change as I thought it would be.

bzbarsky-apple reviewed (dismissed):
Marking changes requested so this does not get accidentally merged until @sagar-apple has a chance to take a look in early January; he is worried about this breaking existing commissioning bits, but hasn't had a chance to look before winter break interfered....

If this gets to a point where this changes requested is the only thing blocking it from landing and it's after Jan 3, please ping me on Slack.

cecille committed:
I'm dumb.

cecille committed:
Merge remote-tracking branch 'upstream/master' into mergy_mergy

cecille committed:
Error checking on parameter setting.

cecille committed:
Fix sizing on thread data set

cecille committed:
Merge remote-tracking branch 'upstream/master' into all_networking_rebased

cecille committed:
Use Variant instead of union.

cecille committed:
Simplify error handling.

cecille committed:
Merge remote-tracking branch 'upstream/master' into all_networking_rebased

pan-apple reviewed (approved):
It looks good to me. The code is moved within the controller API, from Controller.cpp and CommissioneeDeviceProxy to AutoCommissioner, and CommissioningDelegate.

It does change the state machine that drives the commissioning process. So some testing for all use-cases will be helpful. A lot of the use cases are tested via CI. There are some controller/commissioner apps that are not in CI.

cecille committed:
Merge remote-tracking branch 'upstream/master' into all_networking_rebased

cecille committed:
Get rid of some debug cruft.

andy31415 commented:
fast track: this has been up for quite a while now and has had several review  passes.

cecille committed:
Merge remote-tracking branch 'upstream/master' into all_networking_rebased

tcarmelveilleux commented on src/controller/CHIPDeviceController.h:
```diff
@@ -564,6 +564,21 @@ class DLL_EXPORT DeviceCommissioner : public DeviceController,
+     * @param[in] dac                 DAC certificates
+     * @param[in] proxy               device proxy that is being attested.
+     */
+    CHIP_ERROR ValidateAttestationInfo(const ByteSpan & attestationElements, const ByteSpan & signature,
+                                       const ByteSpan & attestationNonce, const ByteSpan & pai, const ByteSpan & dac,
+                                       DeviceProxy * proxy);
```
This will need to be updated to take in more info, such as current vid/pid previously read from basic info cluster. Can be done as a follow-u[

cecille replied:
Basic read will be handled in a follow up. Not sure about the validation. Is there an issue for that open?

msandstedt committed:
cleanup operational ID extraction from certificates

This amendment takes advantage #12915, which makes available
both the root certificate and NOC in memory at the same time
during commissioning.  Because of this, the commissioner no
longer needs to extract and store the Root CA public key to
generate the compressed fabric ID.

cecille committed:
Merge pull request #3 from msandstedt/all_networking_rebased

cleanup operational ID extraction from certificates
